### ISSUE TITLE
provider would not pull the new image when the tag is same

### ISSUE THREAD

**Issue by arno01** [ovrclk/akash#1354]
## Reproducer 

1. build docker image "test:123" which would just "echo 123" and "docker push" it to the registry;
2. create the deployment (send-manifest);
3. get the logs from the provider (provider lease-logs);
4. close the deployment
5. repeat first 4 steps except that do some modification to the image in step 1, say "echo 555", but keep the same image tag "test:123".

## Expected behavior:

Provider should pull the new image.

## Actual behavior:

The provider will not pull the new image, it will start the old image.

## Workaround:

Always change the image tag should there be any update to the docker image.

## Provider in question:

> I haven't tested different providers.

```
"equinix-metal-ams1","akash","mn2-ng","https://provider.provider-0.prod.ams1.akash.pub:8443","akash14c4ng96vdle6tae8r4hc2w4ujwrsh3x9tuudk0"
```

**Comment by arno01** [comment]
cc @dmikey

**Comment by boz** [comment]
@arno01, this is expected behavior.  You need to use a different tag to deploy updates.  Images are heavily cached, pushing changes to a single tag will cause both versions to be running very easily, even if we forced a pull every time.

**Comment by arno01** [comment]
> @arno01, this is expected behavior. You need to use a different tag to deploy updates. Images are heavily cached, pushing changes to a single tag will cause both versions to be running very easily, even if we forced a pull every time.

@boz this should be mentioned in the Akash Docs then, since I've already seen two cases where people would not figure why their image is not working at one provider but does at the other one.

And then this breaks the default Kubernetes behavior for the `:latest` image tag, since it should always be re-pulling images of this tag:
> - omit the imagePullPolicy and use :latest as the tag for the image to use; Kubernetes will set the policy to Always.
>
> Source: https://kubernetes.io/docs/concepts/containers/images/#updating-images

I would actually be more inclined to let people specify in their deployment manifests the [`imagePullPolicy`](https://kubernetes.io/docs/concepts/containers/images/) in order to get more flexible about this.

And then, most of people are going to be using same tags, e.g. `nginx:stable`, `nginx:mainline`, `nginx:latest`, `nginx:1.21` (can replace `nginx` practically with any image).
The issue I see with that (the providers are heavily caching the images) is that the security issues in the libraries of the images  (nor the apps themselves in some cases, i.e. `nginx:stable`, `nginx:1`, `nginx:latest`) will not get pushed as people re-deploy their deployments should they get alerted by [CVE](https://en.wikipedia.org/wiki/Common_Vulnerabilities_and_Exposures) reports.

**Comment by boz** [comment]
it is best practice to use a version tag.  we can add that to the docs.

**Comment by boz** [comment]
It says as much right in your source

> Caution:
You should avoid using the latest tag when deploying containers in production, as it is harder to track which version of the image is running and more difficult to roll back to a working version.

> Instead, specify a meaningful tag such as v1.42.0.

**Comment by arno01** [comment]
@boz 

> It says as much right in your source
> 
> > Caution:
> > You should avoid using the latest tag when deploying containers in production, as it is harder to track which version of the image is running and more difficult to roll back to a working version.
> 
> > Instead, specify a meaningful tag such as v1.42.0.

I agree with that recommendation, however the concern I've raised in this issue has nothing to do with that recommendation.

This is a recommendation for running the production containers which most of people don't.
Many (probably even the majority) are going to be using no tag which defaults to `:latest`.

As right now I'm typing this, there is [3rd occurrence](https://discordapp.com/channels/747885925232672829/771909909335506955/882203392615784518) I'm seeing people are struggling because of this and I'm having to explain them the Akash providers are heavily caching the images, they should use a new tag for every new image update.

My point is that not to lobby for using the `:latest` tag but rather to fix the `imagePullPolicy` which is not currently set to `Always` for the untagged images (or tagged with `:latest` tag) on the Akash providers.
I suggest to stick with the [default imagePullPolicy](https://kubernetes.io/docs/concepts/containers/images/#imagepullpolicy-defaulting)
otherwise there are more and more people will be hitting this issue, there is no doubt about it.

**Comment by tidrolpolelsef** [comment]
My recommendation back when mainnet launched was to block `:latest`, `:stable` etc. The labels don't actually mean anything unfortunately. For example, there is no guaranteed that `:latest` is in fact the latest.

**Comment by andy108369** [comment]
> For example, there is no guaranteed that `:latest` is in fact the latest.

It should guarantee that for the moment someone deploys the app.
It won't mean it's the `:latest` some prolonged time from the moment someone deployed the app.

This is how K8s works and have always been working by default everywhere for the `:latest` tag, I've mentioned that [above](https://github.com/ovrclk/akash/issues/1354#issuecomment-907866119)

I'd prefer sticking with the widely accepted defaults (for `imagePullPolicy`) rather than inventing something new.

And should we want `:latest` to actually **mean** they will be always `:latest`, we should probably look into implementing something like https://keel.sh which would automatically take care updating the images whenever there is a new behind the `:latest` tag. But there should be a toggle for that, i.e. `services.<name>.image_auto_update: true` for example.

As a compromise, the `imagePullPolicy` could be set via the SDL  i.e. `services.<name>.imagePullPolicy: always` for example.

**Comment by andy108369** [comment]
PR https://github.com/ovrclk/provider-services/pull/54

**Comment by andy108369** [comment]
FWIW, I've tested this further, apparently the `HEAD` requests are used for obtaining the image reference and they aren’t triggering the Docker Hub rate limiter upon Pod restart even with the `imagePullPolicy: Always`, as long as image reference haven't been updated on the remote.

 https://asciinema.org/a/541059

The Docker Hub API isn't too restrictive (`100 pulls per 6 hours per IP address`). [Source](https://docs.docker.com/docker-hub/download-rate-limit/)

**Comment by anilmurty** [comment]
Wanted to update this thread after my discussion with @boz today:
Pros of supporting/ allowing users to use the "latest" tag in SDLs (and pulling a fresh uncached image when used or no tag specified):
1. Allows autoupdating of container images in cases like presearch
2. Allows us to potentially build a CI/CD style pipeline where new artifacts get deployed automatically to akash
3. Allows easier integration with solutions like Fleek.co
4. Is consistent with Kubernetes default behavior

For these reasons we are going to work to allow this. The only concern/ downside is making sure we document this behavior clearly and note that a running pod will not be updated unless the pod is terminated (forcing a new pod to be spun up, which will result in the "latest" latest image to be pulled down) - which I believe is how solutions like the presearch "autopdater" handle this anyway (their autoupdater is a fork of https://github.com/containrrr/watchtower)

**Comment by andy108369** [comment]
> forcing a new pod to be spun up, which will result in the "latest" latest image to be pulled down

And this (the pod restart) can be easily triggerred from within the pod itself or the outside (through lease-shell, then `kill <child process of PID 1>` inside the pod).
Either way, this can also be schedule, etc
